crates/llm-local/src/lib.rs

```diff
@@ -0,0 +1,395 @@
+        &mut self,
+        model: wasi_llm::InferencingModel,
+        prompt: String,
+        params: wasi_llm::InferencingParams,
+    ) -> Result<wasi_llm::InferencingResult, wasi_llm::Error> {
+        if !self.allowed_models.contains(&model) {
+            return Err(access_denied_error(&model));
+        }
+        let model = self.inferencing_model(model).await?;
+        let cfg = InferenceSessionConfig {
+            memory_k_type: ModelKVMemoryType::Float16,
+            memory_v_type: ModelKVMemoryType::Float16,
+            n_batch: 8,
+            n_threads: num_cpus::get(),
+        };
+
+        let mut session = Model::start_session(model.as_ref(), cfg);
+        let inference_params = InferenceParameters {
+            sampler: generate_sampler(params),
+        };
+        let mut rng = rand::rngs::StdRng::from_entropy();
+        let mut text = String::new();
+
+        #[cfg(debug_assertions)]
+        {
+            terminal::warn!(
+                "\
+                This is a debug build - running inference might be prohibitively slow\n\
+                You may want to consider switching to the release build"
+            )
+        }
+        let res = session.infer::<Infallible>(
+            model.as_ref(),
+            &mut rng,
+            &llm::InferenceRequest {
+                prompt: prompt.as_str().into(),
+                parameters: &inference_params,
+                play_back_previous_tokens: false,
+                maximum_token_count: Some(params.max_tokens as usize),
+            },
+            &mut Default::default(),
+            |r| {
+                match r {
+                    InferenceResponse::InferredToken(t) => text.push_str(&t),
+                    InferenceResponse::EotToken => return Ok(InferenceFeedback::Halt),
+                    _ => {}
+                };
+                Ok(InferenceFeedback::Continue)
+            },
+        );
+        let stats = res.map_err(|e| {
+            wasi_llm::Error::RuntimeError(format!("Error occurred during inferencing: {e}"))
+        })?;
+        let usage = wasi_llm::InferencingUsage {
+            prompt_token_count: stats.prompt_tokens as u32,
+            generated_token_count: (stats.predict_tokens - stats.prompt_tokens) as u32,
+        };
+        let response = wasi_llm::InferencingResult { text, usage };
+        Ok(response)
+    }
+
+    async fn generate_embeddings(
+        &mut self,
+        model: wasi_llm::EmbeddingModel,
+        data: Vec<String>,
+    ) -> Result<wasi_llm::EmbeddingsResult, wasi_llm::Error> {
+        if !self.allowed_models.contains(&model) {
+            return Err(access_denied_error(&model));
+        }
+        let model = self.embeddings_model(model).await?;
+        generate_embeddings(data, model).await.map_err(|e| {
+            wasi_llm::Error::RuntimeError(format!("Error occurred generating embeddings: {e}"))
+        })
+    }
+}
+
+impl LocalLlmEngine {
+    pub async fn init(&mut self) {
```

Comment:
I don't quite understand why this is needed as a separate step from `new`?

Reply:
It is a step to warm up the cache but yes can move it into `new` if that is a better solution.

Reply:
I would be in favor of that.